File: BinaryTree/php/levelAverages.php
<?php

declare(strict_types=1);

require_once __DIR__ . '/TreeNode.php';

function levelAverages(?TreeNode $root): array
{
    if ($root === null) {
        return [];
    }

    $queue = [$root];
    $averages = [];

    while ($queue) {
        $levelSize = count($queue);
        $sum = 0.0;
        for ($i = 0; $i < $levelSize; $i++) {
            /** @var TreeNode $current */
            $current = array_shift($queue);
            $sum += $current->val;
            if ($current->left !== null) {
                $queue[] = $current->left;
            }
            if ($current->right !== null) {
                $queue[] = $current->right;
            }
        }
        $averages[] = $sum / $levelSize;
    }

    return $averages;
}

File: BinaryTree/csharp/LevelAverages.cs
using System.Collections.Generic;

namespace BinaryTreeSolutions;

public static class LevelAverages
{
    public static IList<double> Solve(TreeNode? root)
    {
        var result = new List<double>();
        if (root is null)
        {
            return result;
        }

        var queue = new Queue<TreeNode>();
        queue.Enqueue(root);
        while (queue.Count > 0)
        {
            var levelSize = queue.Count;
            double sum = 0;
            for (var i = 0; i < levelSize; i++)
            {
                var current = queue.Dequeue();
                sum += current.Val;
                if (current.Left is not null) queue.Enqueue(current.Left);
                if (current.Right is not null) queue.Enqueue(current.Right);
            }
            result.Add(sum / levelSize);
        }

        return result;
    }
}
